### PR TITLE
php80: init at 8.0.0

### PR DESCRIPTION
**Pull request by shyim**
###### Motivation for this change

Amazing new major version of PHP

###### Things done
- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by etu** [comment]
Btw, php-8.0.0 tarballs are now on the mirrors in the expected locations. They haven't released any changelogs yet. But it's worth to start targeting that in this PR.

**Comment by shyim** [comment]
@etu Should I mark all php80Packages and extensions as broken or should I leave them?

**Comment by etu** [comment]
I think it's better to leave them for now, just make sure that the php80 attribute builds with the default set. Then we can test things more and easier when it's cached.

json seems excluded from your default list of extensions, howcome? Is it builtin now or some other reason?

**Comment by shyim** [comment]
json is now always active as extension

**Comment by etu** [comment]
Then we should disable json as extension if it's newer than 8.0 to not cause build errors there :)

**Comment by shyim** [comment]
So I have built 8.0.0 and included the Build Fix for Opcache from https://github.com/php/php-src/commit/4633e70ab1d03cbbfdb166a3c1912d2277c02882 ( I guess will be released with 8.0.1)

@etu I have no glue how to remove json only in php80 attribute 🙈 . Can you help me here?

**Comment by etu** [comment]
> I have no glue how to remove json only in php80 attribute see_no_evil . Can you help me here?

That's easy!
```nix
{ name = "json"; enable = lib.versionOlder php.version "8.0"; }
```

**Comment by shyim** [comment]
Done :)

File: pkgs/top-level/php-packages.nix [hunk 1]
```diff
@@ -388,11 +390,95 @@ in
...
+@@ -31,7 +31,7 @@ if test "$PHP_MM" != "no"; then
+     AC_MSG_ERROR(cannot find mm library)
+   fi
+ 
+-  if test "$enable_zts" = "yes"; then
++  if test "$PHP_THREAD_SAFETY" = "yes"; then
+     dnl The mm library is not thread-safe, and mod_mm.c refuses to compile.
+     AC_MSG_ERROR(--with-mm cannot be combined with --enable-zts)
+   fi
+'')] ++lib.optional (lib.versionOlder php.version "7.4") [
```
Comment (etu) [inline pkgs/top-level/php-packages.nix:481]: I'm thinking that this entire block above should be indented the same way as the block below, the editorconfig build check had opinions about something.

Comment (Ma27) [reply pkgs/top-level/php-packages.nix:481]: What's wrong with adding it into its own file?

Comment (etu) [reply pkgs/top-level/php-packages.nix:481]: Nothing really, it can be it's own file :)

**Comment by etu** [comment]
So now things that depends on the json extension is failing, because it's built in and therefore doesn't exist as a separate extension.

Example in couchbase we have this:
```nix
{
  internalDeps = [ php.extensions.json ];
}
```

Thinking we could do something like:
```nix
{
  internalDeps = [] ++ lib.optionals (lib.versionOlder php.version "8.0") [ php.extensions.json ];
}
```

In places where we use `php.extensions.json`.

**Comment by aanderse** [comment]
I think the `perl` package handles a situation somewhat similar. Without taking the time to think it through what would happen if we set `json = null;` for versions `8.0` and above?

**Comment by etu** [comment]
@aanderse We can start with optionals, we've used it before.

@shyim Could you add this diff:
```diff
diff --git a/pkgs/development/php-packages/couchbase/default.nix b/pkgs/development/php-packages/couchbase/default.nix
index abe4b7efdea..58c276c1a4e 100644
--- a/pkgs/development/php-packages/couchbase/default.nix
+++ b/pkgs/development/php-packages/couchbase/default.nix
@@ -16,7 +16,7 @@ buildPecl {
   configureFlags = [ "--with-couchbase" ];
 
   buildInputs = with pkgs; [ libcouchbase zlib ];
-  internalDeps = [ php.extensions.json ];
+  internalDeps = [] ++ lib.optionals (lib.versionOlder php.version "8.0") [ php.extensions.json ];
   peclDeps = [ php.extensions.igbinary ];
 
   patches = [
diff --git a/pkgs/development/php-packages/redis/default.nix b/pkgs/development/php-packages/redis/default.nix
index 41ff8f38689..a61ae1194ff 100644
--- a/pkgs/development/php-packages/redis/default.nix
+++ b/pkgs/development/php-packages/redis/default.nix
@@ -7,8 +7,9 @@ buildPecl {
   sha256 = "1cfsbxf3q3im0cmalgk76jpz581zr92z03c1viy93jxb53k2vsgl";
 
   internalDeps = with php.extensions; [
-    json
     session
+  ] ++ lib.optionals (lib.versionOlder php.version "8.0") [
+    json
   ] ++ lib.optionals (lib.versionOlder php.version "7.4") [
     hash
   ];
```

This will fix the eval, then I think we can merge this and start testing it.

**Comment by jtojnar** [comment]
For Couchbase, the patch above is not sufficient. It needs https://github.com/couchbase/php-couchbase/pull/31 but that does not apply because we only have legacy branch of php-couchbase. We probably need to remove couchbase from `php80.extensions`.

For Imagick, we need:

```patch
--- a/pkgs/development/php-packages/imagick/default.nix
+++ b/pkgs/development/php-packages/imagick/default.nix
@@ -1,4 +1,4 @@
-{ buildPecl, lib, pkgs, pcre' }:
+{ buildPecl, fetchpatch, lib, pkgs, pcre' }:
 
 buildPecl {
   pname = "imagick";
@@ -6,6 +6,14 @@ buildPecl {
   version = "3.4.4";
   sha256 = "0xvhaqny1v796ywx83w7jyjyd0nrxkxf34w9zi8qc8aw8qbammcd";
 
+  patches = [
+    # Fix compatibility with PHP 8.
+    (fetchpatch {
+      url = "https://github.com/Imagick/imagick/pull/336.patch";
+      sha256 = "nuRdh02qaMx0s/5OzlfWjyYgZG1zgrYnAjsZ/UVIrUM=";
+    })
+  ];
+
   configureFlags = [ "--with-imagick=${pkgs.imagemagick.dev}" ];
   nativeBuildInputs = [ pkgs.pkgconfig ];
   buildInputs = [ pcre' ];

```

**Comment by shyim** [comment]
We could mark couchbase as broken until there is a upstream version of it

**Comment by etu** [comment]
> We could mark couchbase as broken until there is a upstream version of it

Yeah, that's fine :)

**Comment by jtojnar** [comment]
Should we update extensions that have new version supporting PHP 8 here? E.g. [XDebug](http://xdebug.org/announcements/2020-11-25). Even though it might be backwards incompatible

**Comment by shyim** [comment]
@jtojnar XDebug 3 contains also complete new configuration. Maybe build a own package of it?

**Comment by etu** [comment]
xdebug 3.0 is already updated in #105332

**Comment by etu** [comment]
@shyim The optional for json in couchbase is still needed to fix eval

**Comment by shyim** [comment]
Rebased, added json as optional and marked as broken for php 8

**Review by etu — APPROVED** [review]
I'd say that this is good to go for now.

We don't break eval. We have a building package with all default extensions that existed since before in the default install.

We can always touch up things later when it's easier to test with binary caches etc.

Marking packages that doesn't build as broken is a common practice before next release if we don't get to it before then.

**Comment by aanderse** [comment]
:tada: thanks everyone!

**Pull request merged**